### ISSUE TITLE
Upstream master/backup

### ISSUE THREAD

**Issue by fdmota** [erpc/erpc#51]
Does ERPC support upstream selection based on master/backup logic. The idea being that the requests should only be forwarded to the master unless the master is unavailable or as fall behind following the chain head or any other serious issues.
The use case is that the master is a local cheap node and latency is not critical.

**Comment by aramalipoor** [comment]
With the current scoring mechanism only health aspects are considered (latency, rate limits, failures, etc). These scores will be used as a 'weight' when selecting an upstream, so a 'weighted random selection' happens using these scores.

I think your use-case is pretty common that you'd want to give higher weight to a particular upstream. There are few approaches I'm thinking about:
* **(1)** Allow users assign a weight (0.0 to 1.0) which intentionally skews the scores in favor of an upstream.
* or, **(2)** Introduce a concept like 'priority tiers' where you can define 'low' and 'high' tiers for example, and as long as even 1 upstream is is 'alive' stick with the same tier. Definition of 'alive' can be simple or complex (error rate threshold, latency threshold, etc)

Generally 'upstream selection' always brings up these questions:
* When exactly would you consider an upstream broken?
* How often and with what strategy would want to 'give an upstream a chance' to prove themselves?
* Do you sort upstreams and try from best ones? Or do you treat them on/off and completely disable the bad ones?

Basically a trade off of `Cost` (e.g. redundant/hedged calls, cpu/mem pressure, 3rd party provider billing) vs `Performance` (how fast requests are served) and `Resiliency` (how often we finally get a good response).

**Comment by aramalipoor** [comment]
Also have thought about another idea to introduce a `standByRules` config on upstream-level:
```yaml
      - id: my-cheap-node-1
         endpoint: http://my-veeery-cheap-node1.svc.cluster.local
      - id: my-cheap-node-2
         endpoint: http://my-veeery-cheap-node2.svc.cluster.local
      - id: my-cheap-node-3
         endpoint: http://my-veeery-cheap-node2.svc.cluster.local

      - id: my-semi-somewhat-affordable-node
        endpoint: https://www.blastapi.io/xxxxxxxxx
        standByRules:
          - bringAsBackupIf:
              minAvailableUpstreamsLessThan: 3 # bring in if any of cheap ones are down
              errorRateHigherThan: 0.5 # OR if error rates are >=50% on other upstreams
              latencyP90HigherThan: 3s # OR if current upstreams have too high p90 latency

      - id: my-veeeery-expensive-node
        endpoint: https://infura.io/xxxxxxxxx
        standByRules:
          - bringAsBackupIf:
              minAvailableUpstreamsLessThan: 1 # bring in if no upstream is available
              errorRateHigherThan: 0.95 # OR if error rates are >=95% on other upstreams
              latencyP90HigherThan: 10s # OR if current upstreams have too high p90 latency
```

This allows to define these on upstream-level and express that by default these nodes must be standby, and brought on certain conditions.

I'm still thinking about the best way to express this (e.g. maybe on network-level? maybe a more generic idea like 'tier's so we can extend beyond standby use-case, etc).

**Comment by fdmota** [comment]
For me, the rules concept is the best idea for controlling the behavior of how it should forward the request to upstream. I agree it should be generic enough to allow other use cases.
The standby wouldn't be completely idle, right? because the proxy would still have to compare the head of each upstream to unsure that the active upstream is not stuck somewhere.

**Comment by aramalipoor** [comment]
After giving more thoughts introducing inclusion/exclusion rules seems like an interesting idea:
```yaml
- id: my-cheap-node-1
  endpoint: http://my-veeery-cheap-node1.svc.cluster.local
- id: my-cheap-node-2
  endpoint: http://my-veeery-cheap-node2.svc.cluster.local

- id: my-cheap-node-3
  endpoint: http://my-veeery-cheap-node2.svc.cluster.local
  exclusionRules:
    - whenMetrics:
        errorRateHigherThan: 0.95 # OR if error rate is >=95%
        blockHeadLagHigherThan: 1000 # OR if block lag is > 1,000 blocks

- id: my-semi-somewhat-affordable-node
  endpoint: https://www.blastapi.io/xxxxxxxxx
  exclusionRules:
    - excludeByDefault: true
    - whenMetrics:
        errorRateHigherThan: 0.60
  inclusionRules:
    - whenIncludedUpstreams:
        lessThan: 2

- id: my-veeeery-expensive-node
  endpoint: https://infura.io/xxxxxxxxx
  exclusionRules:
    - excludeByDefault: true
    - whenMetrics:
        errorRateHigherThan: 0.60
  inclusionRules:
    - whenIncludedUpstreams:
        lessThan: 1

```

**Comment by aramalipoor** [comment]
> The standby wouldn't be completely idle, right? because the proxy would still have to compare the head of each upstream to unsure that the active upstream is not stuck somewhere.

That's right, the process that determines an upstream's health should be always running in the background, and only inclusion/exclusion of such node must be impacted by this new module.

**Comment by fdmota** [comment]
I think inclusion/exclusion rules cover pretty well situations where there are intermittent failures, but for instance in a situation where the node is completely stuck in time (e.g. not processing new blocks), potentially it would not fail for queries about blocks that it already have data, but that node should not be considered healthy.

**Comment by aramalipoor** [comment]
Implemented in https://github.com/erpc/erpc/pull/69 and is released in 0.0.28 🚀 

Docs will be added after some real-world testing. The PR contains a config example for those who want to test it as 'preview' feature.

By default if you just add "fallback" group to upstreams, then a default policy will be added with a simple logic: https://github.com/erpc/erpc/blob/main/common/defaults.go#L412-L431 you can copy-paste and edit the policy if you want, or provide env vars as shown to control it a bit.
